doc/decisions/error_handling.md

```diff
@@ -0,0 +1,38 @@
+# Error Handling
+
+## Problem
+
+Currently, a lot of functions return the error value, regardless of error. These poses some problems, for instance authors of bindings, where it is hard to tell which error exactly occured. A big improvement to this would be returning different error codes for different types of errors, in order to be able to provide better information in cases of errors. Some functions also make use of an `errorKey`.
+
+While this seems like a decent approach, to provide additional informations in cases of error, it is currently very badly documented. It is not really transparent for the users of the function, what kind of information is included in the `errorKey` and even which messages could possibly be returned.
+
+## Constraints
+
+We also have to consider that many parts of the API very constrained in what values they can use for error codes. If a function returns a pointer the only possible error value by definition is NULL.
+
+## Assumptions
+
+## Considered Alternatives
+
+- Change return types for many functions across the API
+
+## Decision
+
+- Add documentation for errorKey to Doxygen
+- Add second channel for getting information about errors
+- Return additional information about errors from functions where errors can be expected
```

Comment:
```suggestion
- Return error codes directly from functions where failures are expected, e.g. `kdbGet`, `keySetName`
```

To clarify further, IMO this should be the distinction between expected and unexpected failures:
- Expected failures are hard or impossible to detect before calling a function. For example an invalid key name is hard to detect (in general) without calling `keySetName` or a similar function. IO errors during `kdbGet` are impossible to detect beforehand.
- Unexpected failures are errors that can easily prevented by the caller. For example it is very easy to check for `NULL` pointers or locked keys before calling a function.

To be even more concrete, if the check to prevent an error case is almost as expensive as the function call itself (e.g. checking for invalid key names before `keySetName`), then we should return a specific error directly. Depending on the kind of function "return directly" could mean via return value or `errorKey`.